### ISSUE TITLE
Advanced Video Streaming aka HTTP Range Requests

### ISSUE THREAD

**Issue by lastmjs** [dfinity/certified-assets#10]
I'm attempting to do the work to claim this bounty: https://twitter.com/dominic_w/status/1467144071449915395

It seems like implementing HTTP Range request functionality will achieve video streaming, and beyond that audio streaming and really any kind of file streaming. I'm not exactly sure what is in scope for this bounty, but I hope to receive guidance on what is acceptable along the way.

Tentatively:
- [ ]

**Comment by lastmjs** [comment]
An update on my progress. I have 206 partial responses somewhat working. I am now able to load a video player into the default Chrome video view and skip around. 206 http requests are made at any point in the video. The problem is that I now need to stream each partial request, because they could be over the ~3 MB limit. I'm hoping this won't be too hard.

**Comment by paulyoung** [comment]
@lastmjs can you clarify if you are only working on streaming _response_ data, or also on streaming _request_ data?

**Comment by paulyoung** [comment]
I realize that the scope of your work probably relates to the bounty, but it seems like we’ll also need to support streaming uploads as well.

Otherwise, I think the size of the video and audio files involved would be limited by what can be uploaded.

**Comment by paulyoung** [comment]
I believe one limitation here is the 2MB message size.

**Comment by lastmjs** [comment]
Uploads are easily chunked by the client, as can be seen by dfx being able to upload large file sizes

**Comment by paulyoung** [comment]
Right, but that means people need to use `dfx`.

I think the code from the certified assets crate that `dfx` uses would need to be used by `icx-proxy` to enable arbitrary-sized file uploads from other clients in general.

For web-based apps that probably needs changes to the service worker as well.

**Comment by lastmjs** [comment]
What I meant by that is it's easy enough to chunk even from a browser client when uploading. As I understand it the browser client would just need to do what dfx does, which is split the files to be uploaded into appropriate chunks and use the update functions on the asset canister. They wouldn't be able to do it all in one http request, but neither does dfx. And this should be possible on the browser, someone just needs to write the code.

Changes to the asset canister shouldn't be absolutely necessary, but perhaps it would be preferable to implement streaming uploads. But Dfinity is working on increasing the ingress message size limit anyway.

**Comment by paulyoung** [comment]
I’m thinking of the use case where an app is trying to (or limited to) make a regular HTTP upload without any special handling.

It also fits into a category of things that could be moved upstream so that every one who wants to do that doesn’t need to implement the same thing.

**Comment by lastmjs** [comment]
Yeah I agree that would be great. I think for now it might be best to move this functionality into the JS agent, and then arbitrary sized ingress messages I would hope will provide a true fix.

**Comment by lastmjs** [comment]
Just ran into a major issue. I've deployed the canister to production and filled it with assets, but partial responses aren't working in production. I wonder if the boundary nodes are filtering out the response headers.

I added the following response headers:

1. `Accept-Ranges`
2. `Content-Range`

Is there some whitelist in the boundary nodes that needs to be edited?

**Comment by lastmjs** [comment]
More info in this forum post, @jplevyak has said he will talk to the team and see about getting this addressed: https://forum.dfinity.org/t/boundary-node-http-response-headers/10747

**Comment by 3cL1p5e7** [comment]
@lastmjs Hi!)
I am testing adding of custom headers (`Accept-Ranges` and `Content-Range`) and see that headers are received in production

My case
- Hard-code headers in `certified_assets` canister <img width="698" alt="Снимок экрана 2022-02-10 в 01 51 04" src="https://user-images.githubusercontent.com/18343842/153303837-e638f135-e7dc-43b7-a401-f8d65a797b56.png">
 - Deploy and open browser
 - Make `.raw.` request for some asset <img width="1006" alt="Снимок экрана 2022-02-10 в 01 54 06" src="https://user-images.githubusercontent.com/18343842/153304243-33e57141-014a-43e0-bd71-abb965b1e0bc.png">
- Get headers  

Sorry, maybe I misunderstood something in the question, but in my case the headers are returned and it will help :)

**Comment by lastmjs** [comment]
Interesting, I wonder if the 206 status code is not supported then?

**Comment by 3cL1p5e7** [comment]
Sorry, I dont know. I just checked if additional headers are filtered or not for `.raw.` URLs in production and maybe I didn't show anything new

**Comment by 3cL1p5e7** [comment]
I can assume that in order to have a range streaming, needs to upgrade this place
 
https://github.com/dfinity/icx-proxy/blob/6cee84ace323c846f670d2eefdd292aa1b986f08/src/main.rs#L279
At this place, the chunks are downloaded sequentially from the `certified_assets` and possibly sent to the browser. I tried to do something similar in service worker
https://github.com/dfinity/ic/pull/5/files#diff-868be50442ee1d9397fa6910d4627c7a0386365ace603123de7ae169ce2de2e4R220

**Comment by lastmjs** [comment]
@3cL1p5e7 I've done my own testing, and it looks like `206` status codes work in production along with any http response headers I've tried.

It looks like the boundary nodes are filtering out the `Range` http request header in production. I would love to get the boundary nodes to allow `Range` headers, because then I could at least use my fork to have range requests until a more scalable solution is implemented at the boundary node level.

**Comment by 3cL1p5e7** [comment]
@lastmjs 
https://github.com/dfinity/cdk-rs/pull/219

**Comment by lwshang** [comment]
Implemented in #102
